capstone/src/main/java/com/google/sps/Objects/TFIDFStringHelper.java

```diff
@@ -0,0 +1,72 @@
+package com.google.sps.Objects;
+
+import java.util.LinkedHashMap;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A class containing String manipulation helper methods, used when calculating TF-IDF.
+ */
+public class TFIDFStringHelper {
+  
+  /**
+   * "Sanitizes" a String so it can be analyzed.
+   * Remove all punctuation and convert String to lowercase.
+   */
+  public static String sanitize(String input) {
+    if (input ==  null || input.length() == 0) {
+      return input;
+    }
+
+    input = input.toLowerCase();
+    // Regex identifies punctuation except apostrophes.
+    input = input.replaceAll("[\\p{P}&&[^']]", "");
```

Comment:
Sanitization is always fun! I'd be careful of just replacing punctuation with no space, you could be creating new 1-grams this way. Instead I'd replace any you care about with a " " and then take a second pass that turns all 2-or-more-spaces in a row into 1 spaces, and remove all spaces at the beginning and end.

I'd also consider hyphens, dashes, periods, question marks, exclamation points --- do you want to find a valid NGram "here? I was" as "here i was", etc.

Reply:
And then you have crazy cases like users posting urls "https://www.google.com/search?q=term&tbm=shop"